js/setup.js

```diff
@@ -39,12 +39,106 @@ var WIZARD_EYES_COLORS = [
+  '#ee4830',
+  '#30a8ee',
+  '#5ce6c0',
+  '#e848d5',
+  '#e6e848'
+];
+
 var TOTAL_WIZARDS = 4;
 var wizards = [];
 
 var similarList = document.querySelector('.setup-similar-list');
 var template = document.querySelector('#similar-wizard-template').content.querySelector('.setup-similar-item');
 
+var ESC_KEY = 'Escape';
+var ENTER_KEY = 'Enter';
+
+var MIN_NAME_LENGTH = 2;
+var MAX_NAME_LENGTH = 25;
+
+var setup = document.querySelector('.setup');
+var setupOpen = document.querySelector('.setup-open');
+var setupClose = setup.querySelector('.setup-close');
+
+var userNameInput = setup.querySelector('.setup-user-name');
+var coatColorInput = setup.querySelector('input[name=coat-color]');
+var eyesColorInput = setup.querySelector('input[name=eyes-color]');
+var fireballColorInput = setup.querySelector('input[name=fireball-color]');
+
+var wizardCoat = setup.querySelector('.wizard-coat');
+var wizardEyes = setup.querySelector('.wizard-eyes');
+var wizardFireball = setup.querySelector('.setup-fireball-wrap');
+
+// закрытие окна
+var onButtonCloseClick = function () {
+  setup.classList.add('hidden');
+  document.removeEventListener('keydown', onKeydownClick);
+};
+
+// при клике на esc вызывается ф-ция закрытия окна
+var onKeydownClick = function (evt) {
+  if (evt.key === ESC_KEY && document.activeElement !== userNameInput) {
+    onButtonCloseClick();
+  }
+};
+
+// открытие окна
+var onButtonOpenClick = function () {
+  setup.classList.remove('hidden');
+  document.addEventListener('keydown', onKeydownClick);
+};
+
+setupOpen.addEventListener('click', function () {
+  onButtonOpenClick();
+});
+
+setupOpen.addEventListener('keydown', function (evt) {
+  if (evt.key === ENTER_KEY) {
+    onButtonOpenClick();
+  }
+});
+
+setupClose.addEventListener('click', function () {
+  onButtonCloseClick();
+});
+
+setupClose.addEventListener('keydown', function (evt) {
+  if (evt.key === ENTER_KEY) {
+    onButtonCloseClick();
+  }
+});
+
+// проверка валидации формы
+var checkValidity = function () {
+  if (userNameInput.validity.tooShort) {
+    userNameInput.setCustomValidity('Имя должно состоять минимум из 2-х символов');
+  } else if (userNameInput.validity.tooLong) {
+    userNameInput.setCustomValidity('Имя не должно превышать 25-ти символов');
+  } else if (userNameInput.validity.valueMissing) {
+    userNameInput.setCustomValidity('Обязательное поле');
+  } else {
+    userNameInput.setCustomValidity('');
+  }
+};
+
+// проверяет ввод имени пользователя
+var checkInput = function (evt) {
+  var target = evt.target;
+  if (target.value.length < MIN_NAME_LENGTH) {
+    target.setCustomValidity('Имя должно состоять минимум из ' + MIN_NAME_LENGTH + '-х символов');
+  } else if (target.value.length > MAX_NAME_LENGTH) {
+    target.setCustomValidity('Имя должно состоять максимум из ' + MAX_NAME_LENGTH + '-х символов');
+  } else {
+    target.setCustomValidity('');
+  }
+};
+
+userNameInput.addEventListener('invalid', checkValidity);
```

Comment:
```suggestion
userNameInput.addEventListener('invalid', onUserNameInputInvalid);
```
я вообще спец по названиям функций обработчиц событий 😎